Correctly apply defaults to mapped task flow

uranusjr opened the pull request:
Since a taskflow @task is mapped differently from classic operator
classes, we missed applying DAG and task group level arguments to it.
This adds that (and a couple of tests for make sure it sticks).

A small refactoring is also introduced to merge the two default
extraction functions into one because we don't really need two of them;
they are always used together.

uranusjr committed:
Correctly apply defaults to mapped task flow

Since a taskflow @task is mapped differently from classic operator
classes, we missed applying DAG and task group level arguments to it.
This adds that (and a couple of tests for make sure it sticks).

A small refactoring is also introduced to merge the two default
extraction functions into one because we don't really need two of them;
they are always used together.

ashb commented on airflow/decorators/base.py:
```diff
@@ -305,13 +311,21 @@ def expand(self, **map_kwargs: "Mappable") -> XComArg:
+            dag=dag,
+            task_group=task_group,
+            task_params=task_kwargs.pop("params", None),
+            task_default_args=task_kwargs.pop("default_args", None),
```
`default_args` isn't a valid argument to Operators is it?

uranusjr replied:
It actually is, there’s a `default_args` argument on BaseOperator.

ashb replied:
Oh well that's just weird. I wonder why it's there.

ashb commented on tests/models/test_baseoperator.py:
```diff
@@ -877,3 +877,30 @@ def test_expand_mapped_task_instance_skipped_on_zero(dag_maker, session):
+        MockOperator.partial(task_id="mapped").expand(arg1=[1], arg2=[2, 3])
+
+    assert dag.get_task("simple").sla == timedelta(minutes=30)
+    assert dag.get_task("mapped").sla == timedelta(minutes=30)
```
We are about to make SLA not work with mapped operators so

```suggestion
    with dag_maker(default_args={"execution_timeout": timedelta(minutes=30)}) as dag:
        MockOperator(task_id="simple", arg1=None, arg2=0)
        MockOperator.partial(task_id="mapped").expand(arg1=[1], arg2=[2, 3])

    assert dag.get_task("simple").execution_timeout == timedelta(minutes=30)
    assert dag.get_task("mapped").execution_timeout == timedelta(minutes=30)
```

ashb commented on tests/models/test_baseoperator.py:
```diff
@@ -877,3 +877,30 @@ def test_expand_mapped_task_instance_skipped_on_zero(dag_maker, session):
+
+
+def test_mapped_task_applies_default_args_taskflow(dag_maker):
+    with dag_maker(default_args={"sla": timedelta(minutes=30)}) as dag:
```
```suggestion
    with dag_maker(default_args={"execution_timeout": timedelta(minutes=30)}) as dag:
```

ashb commented on tests/models/test_baseoperator.py:
```diff
@@ -877,3 +877,30 @@ def test_expand_mapped_task_instance_skipped_on_zero(dag_maker, session):
+        mapped.expand(arg=[1, 2])
+
+    assert dag.get_task("simple").sla == timedelta(minutes=30)
+    assert dag.get_task("mapped").sla == timedelta(minutes=30)
```
```suggestion
    assert dag.get_task("simple").execution_timeout == timedelta(minutes=30)
    assert dag.get_task("mapped").execution_timeout == timedelta(minutes=30)
```

ashb committed:
Use execution_timeout in mapped operator tests

We are going to make sla not work in mapped operators, so don't use it
in tests.
